enhancements/upstream-downstream-split.md

```diff
@@ -0,0 +1,228 @@
+
+The `operator-framework/operator-lifecycle-manager`, `operator-framework/operator-registry` and `operator-framework/api` repositories will serve as the upstream repositories while a new repository `openshift/operator-lifecycle-manager` will be the downstream monorepo for all of the listed upstream repositories (and any that may be added in future).
+
+All downstream artifacts are built from a new downstream monorepo. Upstream operator-framework repositories are staged as git subtrees in a /staging directory, and corresponding “replace” directives exist in go.mod in order to resolve upstream dependencies from the copy in /staging, e.g.: “replace github.com/operator-framework/operator-lifecycle-manager => ./staging/operator-lifecycle-manager”.
+
+To minimize divergence we will have staging branches composed of upstream releases + a set of patches that are planned to be introduced upstream. If patches do not make it into the next upstream release, they will be cherry-picked into the next staging branch.
+
+### Downstreaming
+1. Change pushed to upstream/x/master
+2. Nightly job syncs upstream master with downstream using an [`UPSTREAM_MERGE.sh`](https://github.com/jmrodri/scripts/blob/master/UPSTREAM-MERGE.sh#L69) script for each subtree
+``` bash
+# UPSTREAM_MERGE.sh snippet
+if ! git diff-index HEAD --exit-code --quiet 2>&1 ; then
+	die "Working tree has modifications.  Cannot add."
+fi
+if ! git diff-index --cached HEAD --exit-code --quiet 2>&1 ; then
+	die "Index has modifications.  Cannot add."
+fi
+
+ref="${3:-master}"
+
+# check state of working directory
+git diff-index --quiet HEAD || { printf "!! Git status not clean, aborting !!\\n\\n%s" "$(git status)"; exit 1; }
+
+# update remote
+git fetch -t "$upstream_remote" "$ref"
+
+rev=$(git rev-parse $default --revs-only FETCH_HEAD) || exit $?
+
+if test "${#rev}" -ne 1; then
+	die "Multiple revisions found at HEAD: Expecting one, Got: '$rev'"
+fi
+
+git merge --no-commit -Xsubtree=$subtree_dir $rev
+
+# unmerged files are overwritten with the upstream copy
+unmerged_files=$(git diff --name-only --diff-filter=U --exit-code)
+differences=$?
+if [[ $differences -eq 1 ]]; then
+  unmerged_files_oneline=$(echo "$unmerged_files" | paste -s -d ' ')
+  git checkout --theirs -- $unmerged_files_oneline
+  if [[ $(git diff --check) ]]; then
+    echo "All conflict markers should have been taken care of, aborting."
+    exit 1
+  fi
+  git add -- $unmerged_files_oneline
+else
+  unmerged_files="<NONE>"
+fi
+
+echo "$rev" > "$subtree_dir/UPSTREAM_VERSION"
+git add UPSTREAM-VERSION
+
+# just to make sure an old version merge is not being made
+git diff --staged --quiet && { echo "No changed files in merge?! Aborting."; exit 1; }
+
+git commit -m "Merge upstream ref $ref" -m "Merge executed via ./UPSTREAM-MERGE.sh $upstream_remote $ref" -m "$(printf "Overwritten conflicts:\\n%s" "$unmerged_files")"
+
+```
+3. Carry over patches from downstream/master to new release branch.
```

Comment:
The new release branch is in downstream, but it will still be a monorepo with all the dependencies. Prior to cutting the release branch, any patches that are already present upstream will be removed from the patches directory and the remaining will be carried forward to the new branch. I can add more detail into this process.